prediction_market_agent/agents/microchain_agent/microchain_agent.py

```diff
@@ -80,8 +83,18 @@ def build_agent(
         engine.register(f)
 
     agent.max_tries = 3
-    print(system_prompt)
-    agent.system_prompt = system_prompt.format(engine_help=engine.help)
+
+    # Restore the prompt from a historical session, replacing the editable part with it.
+    if prompt_handler:
+        if historical_prompt := prompt_handler.fetch_latest_prompt():
```

Comment:
In PromptHandler's `save_prompt` function, it's accessing session identifier from `self`:

```
def save_prompt(self, prompt: str) -> None:
        """Save item to storage."""
        prompt_to_save = Prompt(
            prompt=prompt,
            datetime_=utcnow(),
            session_identifier=self.session_identifier
            if self.session_identifier
            else PROMPT_DEFAULT_SESSION_IDENTIFIER,
        )
        self.storage.save_multiple([prompt_to_save])
```

However, in the `fetch_latest_prompt` method, session identifier is expected as an argument:

```
def fetch_latest_prompt(
        self, session_identifier: str = PROMPT_DEFAULT_SESSION_IDENTIFIER
    ) -> Prompt | None:
        return self.storage.load_latest_prompt(session_identifier=session_identifier)
```

So here at this line, it always fetches prompt for the `PROMPT_DEFAULT_SESSION_IDENTIFIER`, instead of the actual session identifier in PromptHandler.

Btw that's why I suggested before that the session identifier shouldn't be a default argument because it's easy to miss a bug like this.

Reply:
In the comments below
https://github.com/gnosis/prediction-market-agent/pull/238#discussion_r1642432944
https://github.com/gnosis/prediction-market-agent/pull/238#discussion_r1642434410

you said
> Why not just make it a default argument in the function definition? But should there be a default value?

So I'm not 100% sure we both agree on the implementation. Happy to refactor this down the line and discuss more - for the time being:
-> I refactored the function `fetch_latest_prompt` to use the `session_identifier` from the instance if that's not `None`, else use the default value. I think that's what your suggestion was.
-> For the future, my original idea (which I still think is good) is to have `None` as session_identifier all the way until the `Prompt` gets written, and the DB has a default value of PROMPT_DEFAULT_SESSION_IDENTIFIER. That is also used when fetching things. This avoids default arguments being passed multiple times and bugs as you discovered above.